### PR TITLE
Add Silent Ingest for Group Chat Messages

### LINKED ISSUE #15268: [Feature] Native listen-only mode: receive all messages without LLM invocation
## Summary

OpenClaw needs a native "listen-only" or "silent ingest" mode for group chats where the agent should **see all messages** (for context/indexing) but **only invoke the LLM when explicitly mentioned or asked a direct question**.

Currently, the only way to achieve this is `requireMention: false` + a `systemPrompt` instructing the agent to reply `NO_REPLY` for non-relevant messages. This wastes tokens on every single message just to produce a "stay silent" response.

## Problem

With `requireMention: false`, every message in the group triggers a full LLM call. For a busy group with 50+ messages/day, this means:
- **Significant token waste** — each "NO_REPLY" decision costs input tokens (system prompt + context) + output tokens
- **Unnecessary API calls** — hitting rate limits faster
- **Higher latency** for actual requests — queue congestion from silent processing

## Proposed Solution

Add native config-level support for selective response:

```json
{
  "channels": {
    "telegram": {
      "groups": {
        "-123456789": {
          "listenMode": "all",
          "respondMode": "mention",
          "silentIngest": {
            "enabled": true,
            "hooks": ["session-memory"]
          }
        }
      }
    }
  }
}
```

### Key concepts:

1. **`listenMode: "all"`** — Agent receives all messages into session context (current `requireMention: false` behavior)

2. **`respondMode: "mention" | "all" | "allowFrom"`** — Controls when the LLM is actually invoked:
   - `"mention"` — only when @mentioned or replied to
   - `"all"` — every message (current default when `requireMention: false`)
   - `"allowFrom"` — only for messages from specific sender IDs

3. **`silentIngest`** — Messages that don't trigger LLM invocation are still:
   - Added to session context (so agent has full history when invoked)
   - Processed by configured hooks (e.g., `session-memory` for indexing)
   - NOT sent to the LLM API

### Additional enhancement: Per-sender response permissions

Currently `allowFrom` is binary — a sender either triggers the agent or is ignored entirely. A more granular model would allow:

```json
{
  "allowFrom": {
    "293894843": { "canCommand": true, "canTriggerResponse": true },
    "355322273": { "canCommand": true, "canTriggerResponse": true },
    "999999999": { "canTriggerResponse": false }
  }
}
```

This enables scenarios like "team members can command the bot, but client messages are only observed."

## Use Cases

1. **Client observation** — Bot in client chat, listens to everything, only responds when team @mentions it
2. **Internal work chat** — Bot indexes all discussions into knowledge graph, responds only when asked
3. **Meeting notes** — Bot captures all messages but doesn't interrupt conversation flow
4. **Cost optimization** — Reduce API calls by 80-90% in busy groups

## Current Workaround

```json
{
  "requireMention": false,
  "systemPrompt": "You receive ALL messages. Only respond when @mentioned or asked directly. For ALL other messages: reply with exactly NO_REPLY."
}
```

This works but wastes tokens on every message and relies on prompt compliance (soft enforcement).

## Environment
- OpenClaw 2026.2.9 (stable)
- Telegram groups with mixed listen/respond needs

---

## Status update (2026-02-17)
Implemented in #15841 (open PR).

The delivered design is more generalized than the initial request:
- shared silent-ingest pipeline
- shared group ingest policy resolver
- explicit `message_ingest` hook contract
- Telegram + Signal config/schema support

This aligns with current OpenClaw architecture (centralized policy resolution + thin channel adapters + typed hook contracts).

### PR DESCRIPTION
**Pull request by napetrov**
# Add Silent Ingest for Mention-Gated Group Chats

Fixes #15268

## What this PR does

Implements a native **silent ingest** path for group messages that are skipped by mention-gating (`requireMention: true`).

When a non-mentioned group message is skipped:
- no LLM run is started,
- no tools are invoked by core auto-reply,
- but plugins can still observe the message through a dedicated `message_ingest` hook.

This removes the old "NO_REPLY prompt workaround" cost pattern and makes the behavior explicit in config/runtime.

## Why this is now more generalized (not channel-specific glue)

This PR started as channel-local logic and was refactored into a shared pipeline:

- `runSilentMessageIngest(...)` shared helper with common guardrails
- shared group-policy resolver `resolveChannelGroupIngest(...)`
- dedicated hook contract `message_ingest` in plugin types/runner
- channel adapters (Telegram/Signal) only map inbound context into the shared path

So the model is now **platform-aligned** and extendable to other group-capable channels without re-implementing safety logic.

## Design alignment with current OpenClaw architecture

The implementation follows existing OpenClaw patterns:
- policy resolution in `config/group-policy.ts` (group-specific > wildcard default)
- channel handlers stay thin and delegate cross-cutting behavior to shared helpers
- explicit hook typing + wiring in plugin runner (`message_ingest` vs overloading `message_received`)
- fail-open operational behavior (ingest failures/timeouts do not block message routing)

## Key behavior and safety guarantees

- Ingest runs only on mention-skip path for group messages.
- `runSilentMessageIngest` enforces:
  - empty-content short-circuit,
  - `hasHooks("message_ingest")` short-circuit,
  - per-conversation + global inflight caps,
  - timeout wait cap,
  - inflight slot retention until hook settles (prevents timeout over-admission),
  - sender sanitization for hook event `from`.
- Telegram topic isolation uses conversation key `chatId:threadId`.

### Telegram
- `TelegramGroupConfig.ingest?: boolean`
- `TelegramTopicConfig.ingest?: boolean`
- Zod schema updated for both group and topic.

### Signal
- `SignalAccountConfig.groups?: Record<string, SignalGroupConfig>`
- `SignalGroupConfig.ingest?: boolean`
- Zod schema updated to accept `channels.signal.groups.*.ingest`.

## Hook contract

Adds `message_ingest` hook to plugin hook names/handler map and exposes `runMessageIngest(...)` in hook runner.

Semantics:
- passive ingest event,
- no automatic LLM/tool execution in this path,
- async non-blocking from channel routing perspective.

## Tests added

- `src/channels/silent-ingest.test.ts`
  - sanitization path
  - no-hook short-circuit
  - timeout timer cleanup
  - inflight retention under timeout
- `src/config/group-policy.ingest.test.ts`
  - group vs wildcard precedence
- `src/config/zod-schema.providers-core.telegram-ingest.test.ts`
- `src/config/zod-schema.providers-core.signal-ingest.test.ts`
- `src/utils/sanitize.test.ts`

## Practical impact

For busy group chats, this enables memory/indexing hooks on non-mentioned traffic **without paying per-message LLM cost**, while keeping responses mention-gated and consistent with existing group policy behavior.

<h3>Greptile Summary</h3>

This PR adds a **silent ingest path** for group messages that are skipped by mention-gating. When `requireMention: true` and the message contains no mention, the new `message_ingest` hook fires asynchronously so plugins can observe the traffic (e.g. for memory/indexing) without paying LLM cost. The implementation is well-structured: a shared `runSilentMessageIngest` helper enforces per-conversation and global inflight caps, a configurable timeout, and sanitization of the `from` field; both the Telegram and Signal adapters delegate to it with minimal adapter-level glue.

Key points from the review:

- **Overall design is sound.** The shared helper, policy resolver, and hook type wiring all follow existing patterns in the codebase. The inflight/timeout logic is correct and the test for background-hook inflight retention is a nice edge-case catch.
- **`sanitizeUserText` output can exceed `maxLength` by 3 characters** — the `"..."` suffix is appended *after* slicing to `maxLength`, so the result is `maxLength + 3` chars. This is a minor style issue, but the docstring says "limits length" which implies a strict cap.
- **Message content is not sanitized for control characters.** Only the `from` field passes through `sanitizeUserText`; `content` receives only `.trim()`. Control characters (null bytes, embedded newlines, etc.) in the message body will reach `message_ingest` plugin handlers unchanged. This appears to be an intentional design choice (the PR description explicitly calls out "sender sanitization"), but it is worth a second look for log-injection safety in plugins.
- **`enabled` parameter is always `true` at both call sites.** Each caller guards with `if (ingestEnabled)` before calling `runSilentMessageIngest({ enabled: true, ... })`, making the `enabled` parameter inside the function redundant. This is cosmetic but could cause confusion about where the guard lives.

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with minor style concerns; no logic or security bugs that would break existing behavior.
- The core inflight/timeout logic is correct, the hook wiring is clean, and the change is additive (off by default, requires explicit `ingest: true` in config). The three findings are all style-level: (1) `sanitizeUserText` outputs up to `maxLength + 3` chars, (2) message content isn't control-char sanitized before reaching plugins, and (3) the `enabled` param is always hardcoded to `true` at call sites. None of these break runtime behavior today.
- src/utils/sanitize.ts (truncation length contract), src/channels/silent-ingest.ts (content sanitization gap and redundant `enabled` param)

<sub>Last reviewed commit: fd7ed0a</sub>

### CODE REVIEW

**Comment by napetrov** [comment]
Fixed all code review issues:

**P0 - Malformed import (line 58):** ✅ Fixed
- Moved `getGlobalHookRunner` import outside of the `./bot/helpers` import block

**P1 - Hook filtering not enforced (line 504/527):** ✅ Fixed with dedicated hook type
- Added new `message_ingest` hook type to plugin system
- Plugins must explicitly register for `message_ingest` (not `message_received`)
- This provides true security boundary - no accidental execution of wrong hooks
- Removed `ALLOWED_INGEST_HOOKS` whitelist (no longer needed)

**P1 - Zod schema missing (line 160):** ✅ Fixed by simplifying config
- Changed `ingest` from `{enabled: boolean, hooks: string[]}` to just `boolean`
- Hook type itself determines what runs (not config array)
- Simpler API, better type safety

**Config Changes:**
Before:
```yaml
ingest:
  enabled: true
  hooks: ["session-memory"]
```

After:
```yaml
ingest: true  # Runs all message_ingest hooks
```

Changes pushed in commit 7706dc71c

File: src/telegram/bot-message-context.ts [hunk 1]
```diff
@@ -485,6 +486,67 @@ export const buildTelegramMessageContext = async ({
...
+        
+        if (enabled && hooks.length > 0 && rawBody && rawBody.trim().length > 0) {
+          const hookRunner = getGlobalHookRunner();
+          if (hookRunner) {
+            const { ALLOWED_INGEST_HOOKS } = await import("../config/ingest-hooks.js");
+            const { sanitizeUserText } = await import("../utils/sanitize.js");
+            
+            const validHooks = hooks.filter((h): h is string => 
+              typeof h === 'string' && ALLOWED_INGEST_HOOKS.includes(h as any)
+            );
```
[the remark this answers is not in the document]
Comment (napetrov) [reply src/telegram/bot-message-context.ts]: fixed, please rescan

**Comment by napetrov** [comment]
Update: this PR has been refactored into a more generalized design (not channel-local glue).

What changed since earlier drafts:
- Unified shared ingest pipeline via `runSilentMessageIngest(...)`
- Added shared resolver `resolveChannelGroupIngest(...)`
- Added explicit plugin hook contract `message_ingest`
- Wired schema+types for both Telegram and Signal group ingest
- Added inflight/timeout hardening and targeted regression tests

Why this is better:
- clear separation of concerns (channel adapter vs shared policy/safety)
- better alignment with existing OpenClaw policy resolution + hook architecture
- lower long-term maintenance cost and easier extension to more channels
- same mention-gating UX with significantly lower token waste vs NO_REPLY workaround

File: src/utils/sanitize.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  // Trim whitespace
+  const trimmed = chars.join("").trim();
+  if (!trimmed) {
+    return undefined;
+  }
+
+  if (trimmed.length > maxLength) {
+    return `${trimmed.slice(0, maxLength)}...`;
+  }
```
[the remark this answers is not in the document]
Comment (napetrov) [reply src/utils/sanitize.ts]: Fixed in latest head `e9a813639` (commit includes `fix: address ingest review feedback and sanitize length cap`).

Updated truncation budget in `sanitizeUserText` so returned value never exceeds `maxLength`:
- now uses `slice(0, Math.max(0, maxLength - 3)) + "..."`.

File: src/channels/silent-ingest.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  ctx: PluginHookMessageContext;
+  timeoutMs?: number;
+  maxInflight?: number;
+  maxGlobalInflight?: number;
+  hookRunner?: HookRunner | null;
+  log: (message: string) => void;
+  logPrefix: string;
+}): Promise<boolean> {
+  const content = params.event.content?.trim();
+  if (!params.enabled || !content) {
```
[the remark this answers is not in the document]
Comment (napetrov) [reply src/channels/silent-ingest.ts]: Addressed in latest head `e9a813639`.

`runSilentMessageIngest` now sanitizes `content` with `sanitizeUserText(..., 8192)` before dispatching to `message_ingest` hooks, so control characters are stripped consistently (same protection class as `from`).

File: src/channels/silent-ingest.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  enabled: boolean;
+  event: PluginHookMessageReceivedEvent;
+  ctx: PluginHookMessageContext;
+  timeoutMs?: number;
+  maxInflight?: number;
+  maxGlobalInflight?: number;
+  hookRunner?: HookRunner | null;
+  log: (message: string) => void;
+  logPrefix: string;
+}): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (napetrov) [reply src/channels/silent-ingest.ts]: Addressed in latest head `e9a813639`.

Call sites now pass real gating values into the helper (`enabled: ingestEnabled`) instead of hardcoded `true`, and Telegram/Signal wrappers were adjusted so the `enabled` parameter is no longer dead.

File: src/telegram/bot-message-context.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (napetrov) [reply src/telegram/bot-message-context.ts]: Addressed (and now further rebased in latest head `e9a813639`).

We moved ingest onto a dedicated `message_ingest` hook path, so silent ingest no longer dispatches generic `message_received` hooks.
This enforces the intended boundary and avoids executing unrelated hooks.

**Review by napetrov — COMMENTED** [review]
## Review (post-rebase)

### Summary
The design is solid and well-aligned with OpenClaw architecture. The  hook type is the right abstraction — it gives a clean security boundary without needing a whitelist, follows existing policy resolution patterns, and keeps channel adapters thin. The shared  helper avoids code duplication and the inflight/timeout guards are well-reasoned.

### Found & fixed during rebase

**Rebase regression — missing import (P0)**
After rebasing on ,  was no longer imported in  (lost during conflict resolution on an earlier squash commit). This caused  to fail at runtime with `ReferenceError: resolveAccountEntry is not defined`.

Fixed in the rebase tip commit:
```
fix(ingest): add missing resolveAccountEntry import after rebase
```
All 11 ingest tests pass on rebased branch.

---

### Remaining minor issues (style, not blockers)

**1. `sanitizeUserText` length cap off-by-3**
`sanitize.ts` slices to `maxLength` then appends `"..."`, so the output can be `maxLength + 3` chars. Not a security issue but misleads callers who rely on a strict cap:
```ts
// current
if (sanitized.length > maxLength) {
  sanitized = sanitized.substring(0, maxLength) + '...';
}
// fix
if (sanitized.length > maxLength) {
  sanitized = sanitized.substring(0, maxLength - 3) + '...';
}
```

**2. Redundant `enabled: true` param in call sites**
Both Telegram and Signal guard with `if (ingestEnabled)` before calling `runSilentMessageIngest({ enabled: true, ... })`. The `enabled` field inside the helper is always `true` and redundant. Consider removing it from the struct or asserting it to make intent explicit.

**3. Message content not sanitized for control chars**
Only the `from` field is passed through `sanitizeUserText`; raw `content` reaches `message_ingest` plugin handlers as-is. The PR description explicitly scopes sanitization to sender metadata, so this appears intentional — but worth a comment in `runSilentMessageIngest` to make that choice visible for plugin authors.

---

### Verdict

✅ **Approve with minor notes** — no logic or security bugs, design is clean, tests pass post-rebase. The three items above are style/clarity level and can be addressed in a follow-up if desired.

**Review by napetrov — COMMENTED** [review]
## Review (post-rebase)

### Summary
The design is solid and well-aligned with OpenClaw architecture. The `message_ingest` hook type is the right abstraction — clean security boundary without a whitelist, follows existing policy resolution patterns, keeps channel adapters thin. The shared `runSilentMessageIngest` helper avoids duplication and the inflight/timeout guards are well-reasoned.

### Found & fixed during rebase

**Rebase regression — missing import (P0)**
After rebasing on `main`, `resolveAccountEntry` was dropped from imports in `src/config/group-policy.ts` (lost during conflict resolution). This caused `group-policy.ingest.test.ts` to fail at runtime with `ReferenceError: resolveAccountEntry is not defined`.

Fixed in commit on the rebased branch:
```
fix(ingest): add missing resolveAccountEntry import after rebase
```
All 11 ingest tests pass on rebased branch. ✅

---

### Remaining minor issues (style, not blockers)

**1. `sanitizeUserText` length cap off-by-3**
Slices to `maxLength` then appends `"..."`, so output can be `maxLength + 3` chars. Not a security issue but misleads callers expecting a strict cap.

**2. Redundant `enabled: true` param in call sites**
Both Telegram and Signal guard with `if (ingestEnabled)` before calling `runSilentMessageIngest({ enabled: true, ... })`. The `enabled` field inside the helper is always `true` — redundant. Remove or assert.

**3. Message content not sanitized for control chars**
Only `from` goes through `sanitizeUserText`; raw `content` reaches `message_ingest` handlers unchanged. Appears intentional per the PR description, but worth a comment in `runSilentMessageIngest` for plugin authors.

---

### Verdict

✅ **Approve with minor notes** — no logic or security bugs, design is clean, tests pass post-rebase. The three items above are style/clarity level and can be addressed in a follow-up if desired.

File: src/channels/silent-ingest.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  const content = sanitizeUserText(params.event.content, DEFAULT_CONTENT_MAX_LENGTH);
+  if (!params.enabled || !content) {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (napetrov) [reply src/channels/silent-ingest.ts]: Great catch — fixed in 1a43fcabb. I moved the `enabled` guard ahead of `sanitizeUserText`, so disabled ingest now returns immediately without content sanitization work.

**Review by obviyus — CHANGES_REQUESTED** [review]
> if you enable silent ingest for session-memory or command-logger, those handlers will run on skipped group messages

but right now, it sends the event into the plugin message_received pipeline instead, so the two named handlers never run, no?

also, the PR added `ingest` to TypeScript/runtime types, but not to the strict config schema, so `ingest` keys would get rejected as invalid config and won't run.

**Comment by obviyus** [comment]
Closing as duplicate; this was superseded by #60018.
